### ISSUE TITLE
Work around problems introduced by the File Locking app

### ISSUE THREAD

**Issue by oparoz** [owncloud/gallery#54]
The File Locking app is slowing things down a lot: https://github.com/owncloud/core/issues/14107 , but I found that it could also sometimes prevent the app from getting large previews.

**Comment by setnes** [comment]
I am able to reproduce this problem.  

It appears as if the browser window size affects this somehow.  I don't seem to get a broken image unless the window size is larger than the image resolution prior to scaling.  When the window size is smaller, the images load OK even with file locking turned on.

What are the dangers of shutting off the file locking app?

**Comment by setnes** [comment]
After more experiments, I've found it has to be the second fetch for a given size to result in a broken image.  Knowing this I can describe the tests to reproduce it 100% of the time (at least on my box).

1)  Make sure you have some images in a directory that are wider and/or taller than your browser window and some that are smaller than your window.  Change browser window size if necessary.
2)  Make sure file locking app is enabled.
3)  Delete all thumbnails from the user's backend data directory.  (to prove it's the second fetch)
4)  Open the gallery app for the given directory and click through all photos.  (everything should display)
5)  Close the viewer with the X.
6)  Without changing the browser size, use the browser reload button to reload the page.
7)  Click through all the photos.  (large photos display and small photos are broken)

I have preview_max_scale_factor set to 1.  I am not sure what the cause is, but it seems like anything that is scaled to a smaller size doesn't have issues displaying.  Anything that is presented full size will only work the first time for the requested size.  If you change the browser window size and reload.. things work again... until you reload and try again at the same size.

When I do these tests with multiple window sizes, I start to see many identical thumbnails get saved with different names.  This seems related to the images that are smaller than the requested sizes.  They seem unnecessary.

**Comment by oparoz** [comment]
Thanks for debugging this

Did you apply the max preview patch? I'm guessing no because of the thread about the patch.

What seems to be happening is that the Preview class is not able to retrieve the cached preview when file locking is enabled.

I need to see if this is still happening on 8.1...

Regarding all the different cached previews, there is a plan to have categories at some point so that you only have 256x256, 1024x1024 and 2048x2048 pics per example and the browser will get the one bigger than the window size.
But first I'm going to at least fix the fact that there is absolutely no point in creating hundreds of copies of the max preview when the asked size is larger than the max preview size.

**Comment by setnes** [comment]
I have tried this both without the patch to bitmap.php as well as with the latest version of bitmap.php pulled from ownCloud source (has your changes... not sure if it's different than what your patch is trying to get us to).  It didn't matter either way.

I like the idea about static preview sizes.  A simple thing that could be done in the code until then would be to adjust the requested size to the largest multiple of a given number.  Something like...

```
x - ( x mod 20 )  and  y - ( y mod 20 )
```

This would increase the probability that the server has the preview image already and the small gutter created around the image wouldn't be that noticeable.  You'd want to try different multiples.  20 was just a stab in the dark.

**Comment by oparoz** [comment]
I need to find time to play with those settings, but if you want to test your theory:
https://github.com/owncloud/galleryplus/blob/master/js/galleryutility.js#L100

I would use + though and let the browser resize it down.

**Comment by setnes** [comment]
Yeah... if the browser is already downscaling the image for us, then we could use a much larger multiple.  This might eliminate the need for static resizes.  I'll experiment this weekend.

Eliminating the multiple copies of the max preview might require the code to know the image size before requesting a new file.  From what I understand, image metadata is still a future thing for OC.

**Comment by oparoz** [comment]
Eventually having size categories will make things easier because admin will be able to decide what the maximum dimension of a preview should be for various use cases.
We can't arbitrarily send the biggest pic we have because of the time it would take on mobile to get the images.

The max preview magic is done at the PHP level, so we just compare what we're asked to the size of the max preview since there will never be a preview larger than this one.

**Comment by setnes** [comment]
Adding the following after the initialization of width and height seems to work well for me.  It greatly reduces the number of scaled image files created while still being reasonable about the file sizes.  I arbitrarily picked 100 as my multiple.  I'm a little worried about image quality when we're within a few pixels of one of the multiples.  A fudge factor could be added to this for that if needed.  Although being undersized by a few pixels didn't look bad to me, the following will always be larger than the requested size.

```
/* find the next largest image size */
width = width - ( width % 100 ) + 100;
height = height - ( height % 100 ) + 100;
```

This probably could use it's own "issue" thread... although I still feel the fetching of scaled images is very much related to the file locking issue somehow.  There's something different going on if the image file is found to already exist.  Perhaps an overwrite is being attempted when it doesn't need to happen at all.  Where is this code located?

**Comment by oparoz** [comment]
> This probably could use it's own "issue" thread... 

Good idea. Could you create it please?

**Comment by oparoz** [comment]
Found the problem in the files_locking app.
https://github.com/owncloud/files_locking/pull/25/files#diff-079bbe455bfa9aeaf363ae1bc6415ba0R3

**Comment by oparoz** [comment]
The fix only applies to 8.1 since 8.0 doesn't have a preSetup

**Comment by oparoz** [comment]
So, @setnes, if you can test on 8.1. Seems to be fixed there. No patching required.

**Comment by setnes** [comment]
I don't currently have 8.1 installed, so I will have to try that later.  Even if the file locking app has been fixed to not fail, I noticed that something is still recreating and **overwriting** the preview images when the browser size is larger than the image size.  I have confirmed this by watching what happens in the thumbnails directory when file locking is turned off.  This is what is triggering the lock failure and broken image in 8.0.  I'm not sure why the overwrite happens in these cases, and not in cases where the browser size is smaller than the actual image.  I'm new to all of this code.  So even if the locking is fixed, we probably still have an unnecessary preview being created.  (maybe we'll solve it under #157)

**Comment by oparoz** [comment]
I did notice the same thing... There is a write lock when we're trying to retrieve the preview, but I think the file is just "touched" to make sure it's available.

The problem on 8.0 is that the lock is never released and iirc, the info message is talking non-sense.

On 8.1 you get this:
`"PASS: Write lock succeeded on /..."`

**Comment by oparoz** [comment]
I did look at what was happening and it is really overwriting the preview every time...the Preview class needs fixing. I think I'll provide one with Gallery in the future.

**Comment by oparoz** [comment]
Closing this as it's fixed in 8.1.
People should avoid using the file locking app on 8.0.
